neo4j/io/__init__.py

```diff
@@ -955,25 +967,34 @@ def _connect(resolved_address, timeout, keep_alive):
         return s
 
 
-def _secure(s, host, ssl_context):
+def _secure(s, hosts, ssl_context):
     local_port = s.getsockname()[1]
     # Secure the connection if an SSL context has been provided
+    if not hosts:
+        hosts = [None]
     if ssl_context:
-        log.debug("[#%04X]  C: <SECURE> %s", local_port, host)
-        try:
-            sni_host = host if HAS_SNI and host else None
-            s = ssl_context.wrap_socket(s, server_hostname=sni_host)
-        except (SSLError, OSError) as cause:
-            _close_socket(s)
-            error = BoltSecurityError(message="Failed to establish encrypted connection.", address=(host, local_port))
-            error.__cause__ = cause
-            raise error
-        else:
-            # Check that the server provides a certificate
-            der_encoded_server_certificate = s.getpeercert(binary_form=True)
-            if der_encoded_server_certificate is None:
-                s.close()
-                raise BoltProtocolError("When using an encrypted socket, the server should always provide a certificate", address=(host, local_port))
+        last_error = None
+        for host in hosts:
+            log.debug("[#%04X]  C: <SECURE> %s", local_port, host)
+            try:
+                sni_host = host if HAS_SNI and host else None
+                s = ssl_context.wrap_socket(s, server_hostname=sni_host)
+            except (SSLError, CertificateError) as cause:
+                last_error = cause
+                continue
+            except OSError as cause:
+                # No sense in trying another host name with a broken socket
+                last_error = cause
+                break
+            else:
+                # Check that the server provides a certificate
+                der_encoded_server_certificate = s.getpeercert(binary_form=True)
+                if der_encoded_server_certificate is None:
+                    raise BoltProtocolError("When using an encrypted socket, the server should always provide a certificate", address=(host, local_port))
+                return s
+        raise BoltSecurityError(
+            message="Failed to establish encrypted connection.",
+            address=(hosts[0], local_port)) from last_error
```

Comment:
Not needed if `ResolvedAddress.host_names` only contains the host name before DNS resolution.